### PR TITLE
[docs] ControlNet guide

### PR DESCRIPTION
**Pull request by stevhliu**
This PR creates a standalone guide for using ControlNet and ControlNet+SDXL to make the API docs more lightweight and focused as a lookup and reference.

TODO:
- [x] generate and include images from the code snippets
- [x] remove corresponding sections from the API pages

### CODE REVIEW

File: docs/source/en/using-diffusers/controlnet.md [hunk 1]
```diff
@@ -0,0 +1,449 @@
+# ControlNet
+
+ControlNet is a type of model for controlling image diffusion models by conditioning the model with an additional input image. There are many types of conditioning inputs (canny edge, user sketching, human pose, depth, and more) to control a diffusion model. This is hugely useful because it affords you greater control over image generation, making it easier to generate specific images without fiddling with text prompts or denoising values as much.
+
+<Tip>
+
+Check out Section 3.5 of the [ControlNet](https://huggingface.co/papers/2302.05543) paper for a list of ControlNet implementations of various conditioning inputs. You can find all available ControlNet conditioned models on [lllyasviel](https://huggingface.co/lllyasviel)'s Hub profile.
```
Comment (sayakpaul) [inline docs/source/en/using-diffusers/controlnet.md:7]: Users won't be able to find any ControlNet checkpoints for SDXL under their profile, though.

**Comment by sayakpaul** [comment]
@stevhliu which ones you weren't able to run on a free-tier T4 GPU?

File: docs/source/en/using-diffusers/controlnet.md (hunk 1 above)
Comment (sayakpaul) [reply docs/source/en/using-diffusers/controlnet.md:7]: I see. You have linked hf.co/diffusers at the bottom for the SDXL ones. 

Maybe we could also encourage the users to check out the community-trained ControlNets too? 

* SDXL: https://huggingface.co/models?other=stable-diffusion-xl&other=controlnet
* SD: https://huggingface.co/models?other=stable-diffusion&other=controlnet

File: docs/source/en/using-diffusers/controlnet.md [hunk 2]
```diff
@@ -0,0 +1,449 @@
...
+
+Since the locked copy preserves the pretrained model, training and implementing a ControlNet on a new conditioning input is as fast as finetuning any other model because you aren't training the model from scratch.
+
+This guide will show you how to use ControlNet for text-to-image, image-to-image, inpainting, and more! There are many types of ControlNet conditioning inputs to choose from, but in this guide we'll only focus on several of them. Feel free to experiment with other conditioning inputs!
+
+Before you begin, make sure you have the following libraries installed:
+
+```py
+# uncomment to install the necessary libraries in Colab
+#!pip install transformers accelerate safetensors opencv-contrib-python
```
[the remark this answers is not in the document]
Comment (sayakpaul) [reply docs/source/en/using-diffusers/controlnet.md:24]: Also, I think it should `opencv-python`. Contrib modules can safely be excluded here.

File: docs/source/en/using-diffusers/controlnet.md [hunk 3]
```diff
@@ -0,0 +1,449 @@
...
+
+```py
+output = pipe(
+    "spiderman wearing a sombrero", image=canny_image
+).images[0]
+```
+
+## Image-to-image
+
+For image-to-image, you would typically pass an initial image and a prompt to the pipeline to generate a new image. With ControlNet, you can pass an additional conditioning input to steer the model. Let's condition the model with a depth map, an image which contains spatial information. This way, the ControlNet can use the depth map as a control to guide the model to generate an image that preserves spatial information.
```
Comment (sayakpaul) [inline docs/source/en/using-diffusers/controlnet.md:91]: But that is conceptually similar to what we did in the section above, no? We used Canny image to additionally condition the (reverse) diffusion process.

File: docs/source/en/using-diffusers/controlnet.md [hunk 4]
```diff
@@ -0,0 +1,449 @@
...
+    image = depth_estimator(image)["depth"]
+    image = np.array(image)
+    image = image[:, :, None]
+    image = np.concatenate([image, image, image], axis=2)
+    detected_map = torch.from_numpy(image).float() / 255.0
+    depth_map = detected_map.permute(2, 0, 1)
+    return depth_map
+
+depth_estimator = pipeline("depth-estimation")
+depth_map = get_depth_map(image, depth_estimator).unsqueeze(0).half().to("cuda")
```
Comment (sayakpaul) [inline docs/source/en/using-diffusers/controlnet.md:117]: Why do we need to do it like this? 

From the model card [here](https://huggingface.co/lllyasviel/control_v11f1p_sd15_depth), I see there are subtle changes here. I think we can keep things as is here no? We could also potentially reuse the images from the model card too. What do you think?

File: docs/source/en/using-diffusers/controlnet.md [hunk 5]
```diff
@@ -0,0 +1,449 @@
...
+```
+
+Next, load a ControlNet model conditioned on depth maps and pass it to the [`StableDiffusionControlNetImg2ImgPipeline`]. Use the faster [`UniPCMultistepScheduler`] and enable model offloading to speed up inference and reduce memory usage.
+
+```py
+from diffusers import StableDiffusionControlNetImg2ImgPipeline, ControlNetModel, UniPCMultistepScheduler
+import torch
+
+controlnet = ControlNetModel.from_pretrained("lllyasviel/control_v11f1p_sd15_depth", torch_dtype=torch.float16, use_safetensors=True)
+pipe = StableDiffusionControlNetImg2ImgPipeline.from_pretrained(
```
Comment (sayakpaul) [inline docs/source/en/using-diffusers/controlnet.md:127]: Ah I see the pipeline is different. Maybe in the introduction to this section make it bit clearer as to why this pipeline is different from `StableDiffusionControlNetPipeline`.

File: docs/source/en/using-diffusers/controlnet.md [hunk 6]
```diff
@@ -0,0 +1,449 @@
...
+
+Create a function to prepare the control image from the initial and mask images. This'll create a tensor to mark the pixels in `init_image` as masked if the corresponding pixel in `mask_image` is over a certain threshold.
+
+```py
+def make_inpaint_condition(image, image_mask):
+    image = np.array(image.convert("RGB")).astype(np.float32) / 255.0
+    image_mask = np.array(image_mask.convert("L")).astype(np.float32) / 255.0
+
+    assert image.shape[0:1] == image_mask.shape[0:1]
+    image[image_mask > 0.5] = -1.0  # set as masked pixel
```
Comment (sayakpaul) [inline docs/source/en/using-diffusers/controlnet.md:174]: The image pixel range is [0, 1]. Why are we using -1 here?

File: docs/source/en/using-diffusers/controlnet.md [hunk 7]
```diff
@@ -0,0 +1,449 @@
...
+The [`controlnet_conditioning_scale`](https://huggingface.co/docs/diffusers/main/en/api/pipelines/controlnet#diffusers.StableDiffusionControlNetPipeline.__call__.controlnet_conditioning_scale) parameter determines how much weight to assign to the conditioning inputs. A value of 0.5 is recommended for good generalization, but feel free to experiment with this number!
+
+</Tip>
+
+```py
+prompt = "aerial view, a futuristic research complex in a bright foggy jungle, hard lighting"
+negative_prompt = 'low quality, bad quality, sketches'
+
+images = pipe(prompt, negative_prompt=negative_prompt, image=image, controlnet_conditioning_scale=controlnet_conditioning_scale).images[0]
+images
```
Comment (sayakpaul) [inline docs/source/en/using-diffusers/controlnet.md:319]: Could be better formatted (indentation).

File: docs/source/en/using-diffusers/controlnet.md (hunk 3 above)
Comment (stevhliu) [reply docs/source/en/using-diffusers/controlnet.md:91]: Yeah it's similar, but I think it's still valuable for users:

- see an end-to-end example of how to do img2img
- nice to be able to just copy/paste and run the code (they can switch out the images if they want)
- for completeness, it's nice to showcase all the tasks it is capable of

File: docs/source/en/using-diffusers/controlnet.md (hunk 6 above)
Comment (stevhliu) [reply docs/source/en/using-diffusers/controlnet.md:174]: I copied it from [here](https://huggingface.co/docs/diffusers/main/en/api/pipelines/controlnet#diffusers.StableDiffusionControlNetInpaintPipeline.__call__.example), but if that's wrong, I can fix and address that in this PR as well!

Comment (sayakpaul) [reply docs/source/en/using-diffusers/controlnet.md]: Cc: @yiyixuxu here.

File: docs/source/en/api/pipelines/controlnet_sdxl.md [hunk 8]
```diff
@@ -12,149 +12,29 @@ specific language governing permissions and limitations under the License.
...
 
-We provide support using ControlNets with [Stable Diffusion XL](./stable_diffusion/stable_diffusion_xl.md) (SDXL). 
+You can find additional smaller Stable Diffusion XL (SDXL) ControlNet checkpoints from the 🤗 [Diffusers](https://huggingface.co/diffusers) Hub organization, and browse [community-trained](https://huggingface.co/models?other=stable-diffusion-xl&other=controlnet) checkpoints on the Hub.
 
-You can find numerous SDXL ControlNet checkpoints from [this link](https://huggingface.co/models?other=stable-diffusion-xl&other=controlnet). There are some smaller ControlNet checkpoints too:
+<Tip warning={true}>
 
-* [controlnet-canny-sdxl-1.0-small](https://huggingface.co/diffusers/controlnet-canny-sdxl-1.0-small)
```
Comment (patrickvonplaten) [inline docs/source/en/api/pipelines/controlnet_sdxl.md:27]: I'd leave the checkpoints here for now since many non-diffusers controlnet checkpoints pop up when searching for it on the Hub

File: docs/source/en/using-diffusers/controlnet.md [hunk 9]
```diff
@@ -0,0 +1,460 @@
+# ControlNet
```
Comment (patrickvonplaten) [inline docs/source/en/using-diffusers/controlnet.md:1]: Very nice page!

**Review by yiyixuxu — APPROVED** [review]
this is really awesome! super comprehensive

**Review by patrickvonplaten — APPROVED** [review]
Works for me!
